### ISSUE TITLE
Having a really hard time initializing with an LIDF with: ldap_add: Already exists

### ISSUE THREAD

**Issue by mckenziec** [osixia/docker-openldap#438]
Hi. I've spent a good amount of time trying to load a custom LDIF with an Active Directory compatible schema (so I can use OpenLDAP for my work, but I know at some point someone will want me to point to their AD. Boo.) but I've run into a ton of different issues. At this point I have my compose volume mounting a directory with a bootstrap.ldif in it. I want to extend the default LDIF the container creates but it looks like the bootstrap.ldif is run twice? Is it the naming of my ldif file? Or that I'm not setting an env variable?

At the bottom is a clip of the container debug log. You'll see what looks like all the LIDF entries being added, then a failures to add the 1st LDIF item (the Accounting Group) after it already added it?

Thanks for any help.

Here's my compose:

```
  ldap:
    image: osixia/openldap:latest
    container_name: ldap
    domainname: "test.local"
    hostname: "ldap"
    command: --copy-service --loglevel debug
    ports:
      - "389:389"
    env_file:
      - ${PWD}/ldap.env
    volumes:
      - ${PWD}/var/ldap/lib:/var/lib/ldap
      - ${PWD}/var/ldap/etc:/etc/ldap/slapd.d
      - ${PWD}/var/ldap/certs:/container/service/slapd/assets/certs/
      - ${PWD}/conf:/container/service/slapd/assets/config/bootstrap/ldif/custom
    networks:
      - test
```
Here's my ldap.env:

```
LDAP_LOG_LEVEL=256
LDAP_ORGANISATION=Test Inc.
LDAP_DOMAIN=test.local
LDAP_BASE_DN=dc=test,dc=local
LDAP_ADMIN_PASSWORD=admin
LDAP_TLS=false
KEEP_EXISTING_CONFIG=false
LDAP_REMOVE_CONFIG_AFTER_SETUP=true
```

Here's my LDIF (clipped down, it's just 10 users with some groups, generated from the LDIFGen2 tool):

```
version: 1

# groups...
dn: ou=Product Development,dc=test, dc=local
changetype: add
ou: Product Development
objectClass: top
objectClass: organizationalUnit

# users...
dn: cn=Shelby Brady,ou=Product Development,dc=est, dc=local
changetype: add
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
cn: Shelby Brady
sn: Brady
sAMAccountName: BradyS
description: This is Shelby Brady's description
facsimileTelephoneNumber: +1 818 803-4244
l: San Mateo
ou: Product Development
postalAddress: Product Development$San Mateo
telephoneNumber: +1 818 678-5976
title: Supreme Product Development Stooge
userPassword: Password1
uid: BradyS
givenName: Shelby
mail: BradyS@ns-mail8.com
carLicense: HPUM5O
departmentNumber: 2321
employeeType: Contract
homePhone: +1 818 709-7560
initials: S. B.
mobile: +1 818 512-4885
pager: +1 818 235-8610
roomNumber: 8338
manager: null
secretary: null 
```

Here's the debug log from the ldap container:

```
Processing file /container/run/service/slapd/assets/config/bootstrap/ldif/50-bootstrap.ldif
Processing file /container/run/service/slapd/assets/config/bootstrap/ldif/50-bootstrap.ldif
5eeb880d conn=1024 fd=12 ACCEPT from PATH=/var/run/slapd/ldapi (PATH=/var/run/slapd/ldapi)
5eeb880d conn=1024 op=0 BIND dn="" method=163
5eeb880d conn=1024 op=0 BIND authcid="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth" authzid="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth"
5eeb880d conn=1024 op=0 BIND dn="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth" mech=EXTERNAL sasl_ssf=0 ssf=71
5eeb880d conn=1024 op=0 RESULT tag=97 err=0 text=
5eeb880d conn=1024 op=1 ADD dn="ou=Accounting,dc=test,dc=local"
5eeb880d conn=1024 op=1 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=2 ADD dn="ou=Product Development,dc=test,dc=local"
5eeb880d conn=1024 op=2 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=3 ADD dn="ou=Product Testing,dc=test,dc=local"
5eeb880d conn=1024 op=3 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=4 ADD dn="ou=Human Resources,dc=test,dc=local"
5eeb880d conn=1024 op=4 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=5 ADD dn="ou=Payroll,dc=test,dc=local"
5eeb880d conn=1024 op=5 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=6 ADD dn="ou=Janitorial,dc=test,dc=local"
5eeb880d conn=1024 op=6 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=7 ADD dn="ou=Management,dc=test,dc=local"
5eeb880d conn=1024 op=7 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=8 ADD dn="ou=Administrative,dc=test,dc=local"
5eeb880d conn=1024 op=8 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=9 ADD dn="ou=Peons,dc=test,dc=local"
5eeb880d conn=1024 op=9 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=10 ADD dn="ou=Planning,dc=test,dc=local"
5eeb880d conn=1024 op=10 RESULT tag=105 err=0 text=
5eeb880d conn=1024 op=11 ADD dn="cn=Elsi Girard,ou=Payroll,dc=test,dc=local"
5eeb880d conn=1024 op=11 RESULT tag=105 err=17 text=sAMAccountName: attribute type undefined
5eeb880d conn=1024 op=12 UNBIND
5eeb880d conn=1024 fd=12 closed
5eeb880d conn=1025 fd=12 ACCEPT from IP=127.0.0.1:55916 (IP=0.0.0.0:389)
5eeb880d conn=1025 op=0 BIND dn="cn=admin,dc=test,dc=local" method=128
5eeb880d conn=1025 op=0 BIND dn="cn=admin,dc=test,dc=local" mech=SIMPLE ssf=0
5eeb880d conn=1025 op=0 RESULT tag=97 err=0 text=
5eeb880d conn=1025 op=1 ADD dn="ou=Accounting,dc=test,dc=local"
5eeb880d conn=1025 op=1 RESULT tag=105 err=68 text=
5eeb880d conn=1025 op=2 UNBIND
5eeb880d conn=1025 fd=12 closed
ldap_add: Undefined attribute type (17)
	additional info: sAMAccountName: attribute type undefined
adding new entry "ou=Accounting,dc=test, dc=local"

adding new entry "ou=Product Development,dc=test, dc=local"

adding new entry "ou=Product Testing,dc=test, dc=local"

adding new entry "ou=Human Resources,dc=test, dc=local"

adding new entry "ou=Payroll,dc=test, dc=local"

adding new entry "ou=Janitorial,dc=test, dc=local"

adding new entry "ou=Management,dc=test, dc=local"

adding new entry "ou=Administrative,dc=test, dc=local"

adding new entry "ou=Peons,dc=test, dc=local"

adding new entry "ou=Planning,dc=test, dc=local"

adding new entry "cn=Elsi Girard,ou=Payroll,dc=test, dc=local"

ldap_add: Already exists (68)
adding new entry "ou=Accounting,dc=test, dc=local"
*** /container/run/startup/slapd failed with status 68

*** Run commands before finish...
*** Killing all processes...
5eeb880d daemon: shutdown requested and initiated.
5eeb880d slapd shutdown: waiting for 0 operations/tasks to finish
5eeb880d slapd stopped.
```

**Comment by Thyrael** [comment]
I had the same issue and fix it by removing existing entry in my ldif file. In my case, it was admin group that broke my seed. 

Hope that can be usefull in your case

**Comment by mckenziec** [comment]
What I ended up doing is creating my own docker image FROM the osixia openldap image and copying my 50-bootstrap.ldif in to replace the existing one, and included the cn=admin simpleSecurityObject user. Only my ldif is run on initialization.

There's something not being covered in this image that should be worked out and documented. The ability to volume mount an ldif to be run or replaced during 1st time initialization is a great feature, but there are too many ways to screw it up. For example, (if I recall all my experimenting from last week) if you volume mount an ldif file or the directory with your ldif files, a root level operation processing the file in the container will mess up your local host directory. I think it happens when you mix in KEEP_EXISTING_CONFIG and LDAP_REMOVE_CONFIG_AFTER_SETUP. No matter what the combination, its not described in the README how to initialize with your own ldif in relation to the directory/file volume mounting and those settings. I've have to recreate the mount directory a bunch of times trying to experiment. In all cases I had to conclude on just replacing the default bootstrap ldif.

I'm not complaining. I'm just pointing out that such a useful feature (bootstraping the ldif in the container without having to do a post boot step with ldapadd or ldapmodify is really one of the biggest benefits to using a container like this e.g. simplification with consistency) probably needs some attention in the README.

I'm happy with embedding my ldif for now. Thanks for the help Thyrael.
